src/System.Private.CoreLib/src/System/Enum.cs

```diff
@@ -282,262 +280,530 @@ internal static ulong ToUInt64(object value)
+                return true;
             }
 
-            int firstNonWhitespaceIndex = -1;
-            for (int i = 0; i < value.Length; i++)
+            if (failureIsOverflow)
             {
-                if (!char.IsWhiteSpace(value[i]))
-                {
-                    firstNonWhitespaceIndex = i;
-                    break;
-                }
+                if (throwOnFailure) throw new OverflowException(SR.Overflow_UInt64);
             }
-            if (firstNonWhitespaceIndex == -1)
+            else if (TryParseByName(enumType, originalValueString, value, ignoreCase, throwOnFailure, out ulongResult))
             {
-                parseResult.SetFailure(ParseFailureKind.Argument, nameof(SR.Arg_MustContainEnumInfo), null);
-                return false;
+                enumResult = RuntimeHelpers.IsReferenceOrContainsReferences<TResult>() ? (TResult)InternalBoxEnum(enumType, (long)ulongResult) : Unsafe.As<ulong, TResult>(ref ulongResult);
+                return true;
             }
 
-            // We have 2 code paths here. One if they are values else if they are Strings.
-            // values will have the first character as as number or a sign.
-            ulong result = 0;
+            enumResult = default;
+            return false;
+        }
 
-            char firstNonWhitespaceChar = value[firstNonWhitespaceIndex];
-            if (char.IsDigit(firstNonWhitespaceChar) || firstNonWhitespaceChar == '-' || firstNonWhitespaceChar == '+')
+        private static bool TryParseRareEnum<TResult>(RuntimeType enumType, string originalValueString, ReadOnlySpan<char> value, bool ignoreCase, bool throwOnFailure, out TResult enumResult)
+        {
+            if (StartsNumber(value[0]))
             {
                 Type underlyingType = GetUnderlyingType(enumType);
-                object temp;
-
                 try
                 {
-                    value = value.Trim();
-                    temp = Convert.ChangeType(value, underlyingType, CultureInfo.InvariantCulture);
-                    parseResult.parsedEnum = ToObject(enumType, temp);
+                    enumResult = (TResult)ToObject(enumType, Convert.ChangeType(value.ToString(), underlyingType, CultureInfo.InvariantCulture));
                     return true;
                 }
                 catch (FormatException)
-                { // We need to Parse this as a String instead. There are cases
-                  // when you tlbimp enums that can have values of the form "3D".
-                  // Don't fix this code.
+                {
+                    // We need to Parse this as a String instead. There are cases
+                    // when you tlbimp enums that can have values of the form "3D".
                 }
-                catch (Exception ex)
+                catch when (!throwOnFailure)
```

Comment:
Neat. I didn't know about filtering based on a `when` class. TIL (C# 6.0 feature). I am assuming this catches all exceptions when the condition is true? You don't need to specify an exception type explicitly?

Reply:
> I am assuming this catches all exceptions when the condition is true?

Yes.

> You don't need to specify an exception type explicitly?

No.  You can if you want to, e.g.
```C#
catch (MyCoolException e) when (SomeCondition)
{
}
```
but just as you can leave off the exception type without a filter:
```C#
catch
{
}
```
you can do the same with a filter:
```C#
catch when (SomeCondition)
{
}
```